include/trieste/pass.h

```diff
@@ -139,74 +189,35 @@ namespace trieste
         auto start = it;
         for (auto& rule : rules_)
         {
-          match.return_to_frame(0);
-          match.add_frame();
-          if (rule.first.match(it, node->end(), match))
+          match.reset();
+          if (SNMALLOC_UNLIKELY(rule.first.match(it, end, match)))
           {
-            // Replace [start, it) with whatever the rule builds.
-            auto replace = rule.second(match);
-
-            if (replace && (replace == NoChange))
+            replaced = replace(match, rule.second, start, it, node);
+            if (replaced != -1)
             {
-              it = start;
-              continue;
+              changes += replaced;
+              break;
             }
-
-            auto loc = (*start)->location();
-
-            for (auto i = start + 1; i < it; ++i)
-              loc = loc * (*i)->location();
-
-            it = node->erase(start, it);
-
-            // If we return nothing, just remove the matched nodes.
-            if (!replace)
-            {
-              replaced = 0;
-            }
-            else if (replace == Seq)
-            {
-              // Unpack the sequence.
-              std::for_each(replace->begin(), replace->end(), [&](Node n) {
-                n->set_location(loc);
-              });
-
-              replaced = replace->size();
-              it = node->insert(it, replace->begin(), replace->end());
-            }
-            else
-            {
-              // Replace with a single node.
-              replaced = 1;
-              replace->set_location(loc);
-              it = node->insert(it, replace);
-            }
-
-            changes += replaced;
-            break;
-          }
-          else
-          {
-            it = start;
           }
+          it = start;
         }
 
-        if (flag(dir::once))
+        if (replaced == -1)
+        {
+          // If we didn't do anything, advance to the next node.
+          ++it;
+        }
+        else if (flag(dir::once))
         {
           // Skip over everything we populated, or if no change (-1) skip
           // forwards to next term.
-          it += std::abs(replaced);
+          it += replaced;
         }
-        else if (replaced >= 0)
+        else
         {
           // If we did something, reexamine from the beginning.
           it = node->begin();
         }
```

Comment:
```suggestion
```
Removing this branch, which means that it will retry on the thing it has inserted but not back up to the start of the children results in a 7% speedup in rego-cpp.

This is a semantic change, which means patterns like
```
  A * B => C
  D * C => E
  C * F => G
```
applied to 
```
  D A B F
```
would result in
```
  D A B F -> D C F -> D G
```
rather than
```
  D A B F -> D C F -> E F
```

This change does not appear to affect the results of rego-cpp. Not sure if we should add it.